Represent feComponentTransfer CSS filters as feColorMatrix

smfr opened the pull request:
#### d3496a0787a4db8f177e6c4642771fbe7d99c268
<pre>
Represent feComponentTransfer CSS filters as feColorMatrix
<a href="https://bugs.webkit.org/show_bug.cgi?id=245111">https://bugs.webkit.org/show_bug.cgi?id=245111</a>
&lt;rdar://99850797&gt;

Reviewed by NOBODY (OOPS!).

The &quot;feComponentTransfer&quot; type CSS filters (invert, opacity, brightness, contrast) can
be represented as feColorMatrix, which will enable future optimizations.

We were already converting these to color matrix to hand to Core Animation; this change
factors the matrix construction into ColorMatrix.h, and calls that code from CSSFilter
and PlatformCAFiltersCocoa.

Using FEColorMatrix was slower than FEComponentTransfer (e.g. 48ms per invocation vs. 17ms)
but we can make it faster by using vImage when the matrix has values in the last column; we
can represent these as post-bias in the vImage call. This speeds it up (6ms in my test).

* LayoutTests/css3/filters/effect-brightness.html:
* LayoutTests/css3/filters/effect-combined.html:
* LayoutTests/css3/filters/effect-contrast.html:
* LayoutTests/css3/filters/effect-invert.html:
* Source/WebCore/platform/graphics/filters/FEColorMatrix.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::createBrightnessEffect):
(WebCore::createContrastEffect):
(WebCore::createInvertEffect):
(WebCore::createOpacityEffect):
(WebCore::CSSFilter::apply):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3496a0787a4db8f177e6c4642771fbe7d99c268

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22378 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22575 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23456 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24282 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20712 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22632 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26838 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22911 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21720 "Exiting early after 10 failures. 95 tests run. 1 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22617 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22306 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19426 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25137 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19356 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20281 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26535 "Found 1 new test failure: css3/filters/effect-contrast.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20373 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20520 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-backgrounds/border-radius-clipping-with-transform-001.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24389 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/21049 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17850 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-backgrounds/border-radius-clipping-with-transform-001.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20502 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20090 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5336 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-backgrounds/border-radius-clipping-with-transform-001.html (failure)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24739 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/21040 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/07b1536cca4342b47f7ac37384d73ef5138ff98d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/4297

smfr commented:
This is a draft because using FEColorMatrix is about 3x slower in software than FEComponentTransfer in a small test that I made.

weinig commented on Source/WebCore/rendering/CSSFilter.cpp:
```diff
@@ -84,25 +84,30 @@ static RefPtr<FilterEffect> createBlurEffect(const BlurFilterOperation& blurOper
-    ComponentTransferFunction nullFunction;
-    return FEComponentTransfer::create(transferFunction, transferFunction, transferFunction, nullFunction);
+    Vector<float> inputParameters {
+        colorMatrix.at(0, 0), colorMatrix.at(0, 1), colorMatrix.at(0, 2), 0, 0,
```
I think we can probably make a nice helper func on ColorMatrix<> to make these Vectors. (is this row major or column major?).

shallawa replied:
I agree this will be a nice change. This conversion already exists in two places `createGrayScaleEffect()` and `createSepiaEffect()'.  This patch is adding four new ones.

weinig commented on Source/WebCore/platform/graphics/ColorMatrix.h:
```diff
@@ -85,6 +110,33 @@ constexpr ColorMatrix<3, 3> grayscaleColorMatrix(float amount)
+    amount = std::clamp(amount, 0.0f, 1.0f);
+    float multiplier = 1.0f - amount * 2.0f;
+    return ColorMatrix<5, 4> {
+            multiplier, 0.f, 0.f, 0.f, amount,
```
I tried to use 0.0f consistently rather than 0.f.

weinig commented on Source/WebCore/platform/graphics/ColorMatrix.h:
```diff
@@ -85,6 +110,33 @@ constexpr ColorMatrix<3, 3> grayscaleColorMatrix(float amount)
+    return ColorMatrix<5, 4> {
+            multiplier, 0.f, 0.f, 0.f, amount,
+            0.f, multiplier, 0.f, 0.f, amount,
+            0.f, 0.f, multiplier, 0.f, amount,
```
The indentation here doesn't look like it matches the existing matrix helpers.

weinig reviewed (approved):
Very neat

smfr committed:
Represent feComponentTransfer CSS filters as feColorMatrix
https://bugs.webkit.org/show_bug.cgi?id=245111
<rdar://99850797>

Reviewed by NOBODY (OOPS!).

The "feComponentTransfer" type CSS filters (invert, opacity, brightness, contrast) can
be represented as feColorMatrix, which will enable future optimizations.

We were already converting these to color matrix to hand to Core Animation; this change
factors the matrix construction into ColorMatrix.h, and calls that code from CSSFilter
and PlatformCAFiltersCocoa.

Using FEColorMatrix was slower than FEComponentTransfer (e.g. 48ms per invocation vs. 17ms)
but we can make it faster by using vImage when the matrix has values in the last column; we
can represent these as post-bias in the vImage call. This speeds it up (6ms in my test).

* LayoutTests/css3/filters/effect-brightness.html:
* LayoutTests/css3/filters/effect-combined.html:
* LayoutTests/css3/filters/effect-contrast.html:
* LayoutTests/css3/filters/effect-invert.html:
* Source/WebCore/platform/graphics/filters/FEColorMatrix.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::createBrightnessEffect):
(WebCore::createContrastEffect):
(WebCore::createInvertEffect):
(WebCore::createOpacityEffect):
(WebCore::CSSFilter::apply):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/21b71e56dcb0f45c03ebd820a374781a6acc4315. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/4297

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/945c145c419c8c5a431c8cfa3aca9f28e630111c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/4297

shallawa commented on LayoutTests/css3/filters/effect-combined.html:
```diff
@@ -1,7 +1,7 @@
 <html>
 <head>
-    <meta name="fuzzy" content="maxDifference=0-20; totalPixels=0-600" />
+	<meta name="fuzzy" content="maxDifference=0-4; totalPixels=0-5000">
```
tabs -> spaces?

shallawa commented on LayoutTests/css3/filters/effect-contrast-hw.html:
```diff
@@ -1,7 +1,7 @@
 <html>
 <head>
-    <meta name="fuzzy" content="maxDifference=0-230; totalPixels=0-36000">
+	<meta name="fuzzy" content="maxDifference=0-230; totalPixels=0-44000">
```
Ditto.

shallawa commented on LayoutTests/css3/filters/effect-grayscale-hw.html:
```diff
@@ -1,7 +1,7 @@
 <html>
 <head>
-    <meta name="fuzzy" content="maxDifference=0-54; totalPixels=0-33000">
+	<meta name="fuzzy" content="maxDifference=0-45; totalPixels=0-57000">
```
Ditto

shallawa commented on LayoutTests/css3/filters/effect-invert.html:
```diff
@@ -1,7 +1,7 @@
 <html>
 <head>
-    <meta name="fuzzy" content="maxDifference=0-48; totalPixels=0-10000" />
+	<meta name="fuzzy" content="maxDifference=0-48; totalPixels=0-15000">
```
Ditto.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bf7148daee0ff3fa5739f0d0da064b51f0806e97. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/4297

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d3496a0787a4db8f177e6c4642771fbe7d99c268. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/4297

smfr commented:
No longer needed.
